pisa/stages/flux/pi_mceq_barr_red.py

```diff
@@ -0,0 +1,489 @@
+        self.barr_param_names = [
+            # pions
+            "a_f",
+            "g",
+            "h",
+            "i",
+            # kaons
+            "w",
+            "x",
+            "y",
+            "z",
+        ]
+
+        # Define signs for Barr params
+        # +  -> meson production
+        # -  -> antimeson production
+        self.barr_param_signs = ["+", "-"]
+
+        # Atmopshere model params
+        # TODO
+
+        # Get the overall list of params for which we have gradients stored
+        # Define a mapping to index values, will he useful later
+        self.gradient_param_names = [
+            n + s for n in self.barr_param_names for s in self.barr_param_signs
+        ]
+        self.gradient_param_indices = collections.OrderedDict(
+            [(n, i) for i, n in enumerate(self.gradient_param_names)]
+        )
+
+        #
+        # Call stage base class constructor
+        #
+
+        # Define stage parameters
+        expected_params = (
+            # pion
+            "pion_ratio",
+            "barr_a_f_Pi",
+            "barr_g_Pi",
+            "barr_h_Pi",
+            "barr_i_Pi",
+            # kaon
+            "barr_w_K",
+            "barr_x_K",
+            "barr_y_K",
+            "barr_z_K",
+            "barr_w_antiK",
+            "barr_x_antiK",
+            "barr_y_antiK",
+            "barr_z_antiK",
+            # CR
+            "delta_index",
+            "energy_pivot",
+        )
+
+        input_names = ()
+        output_names = ()
+
+        # Using Honda for nominal flux. Keys should already exist
+        input_calc_keys = ("nu_flux_nominal", "nubar_flux_nominal")
+        # what are keys added or altered in the calculation used during apply
+        output_calc_keys = ("nu_flux")
+        # what keys are added or altered for the outputs during apply
+        output_apply_keys = ("nu_flux")
+
+        # store args
+        self.table_file = table_file
+
+        # init base class
+        super(pi_mceq_barr_red, self).__init__(
+            data=data,
+            params=params,
+            expected_params=expected_params,
+            input_names=input_names,
+            output_names=output_names,
+            debug_mode=debug_mode,
+            input_specs=input_specs,
+            calc_specs=calc_specs,
+            output_specs=output_specs,
+            input_calc_keys=input_calc_keys,
+            output_calc_keys=output_calc_keys,
+            output_apply_keys=output_apply_keys,
+        )
+
+        assert self.input_mode is not None
+        assert self.calc_mode is not None
+        assert self.output_mode is not None
+
+    def setup_function(self):
+
+        self.data.data_specs = self.calc_specs
+
+        #
+        # Init arrays
+        #
+
+        # Prepare some array shapes
+        gradient_params_shape = (len(self.gradient_param_names),)
+
+        # Loop over containers
+        for container in self.data:
```

Comment:
probably but I would need to see some example code for this because I"m still learning the pisa ropes here... no idea how easy or hard this change would be..

Reply:
It _should_ work as is with calculations on a grid. PISA takes care of everything under the hood. You can try it out by changing the `calc_specs` in your cfg file to a fine binning.
If that works the we can do the container linking (should just be 3 additional lines in the code)

Reply:
I tried:
`input_specs = events`
`calc_specs = flux_binning`
`output_specs = events`

where the binning is defined as 
`flux_binning.order = true_energy, true_coszen`
`flux_binning.true_energy = {'num_bins':1000, 'is_log':True, 'domain':[0.1, 10000.] * units.GeV, 'tex':r'E_{\rm true}'}`
`flux_binning.true_coszen = {'num_bins':1000, 'is_lin':True, 'domain':[-1,1], 'tex':r'\cos\,\theta_{z,{\rm true}}'}`

 but I get the following error:
`KeyError: '"nu_flux" not found in container "nue_cc"'`

Reply:
can you post some more of the error message? I want to know where it does not find the container

Reply:
[   ERROR] Error occurred computing outputs in stage flux / service pi_mceq_barr_red ...
---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
/afs/ifh.de/group/amanda/scratch/blotsd/Analysis/Oscillations/fork_pisa3/pisa/core/container.py in binned_to_array(self, key)
    394         try:
--> 395             binning, hist = self.binned_data[key]
    396         except KeyError:

KeyError: 'n'

During handling of the above exception, another exception occurred:

ValueError                                Traceback (most recent call last)
<ipython-input-5-4308aadb34d3> in <module>
----> 1 templates = template_maker.get_outputs(return_sum=False)

/afs/ifh.de/group/amanda/scratch/blotsd/Analysis/Oscillations/fork_pisa3/pisa/core/distribution_maker.py in get_outputs(self, return_sum, sum_map_name, sum_map_tex_name, **kwargs)
    200 
    201         """
--> 202         outputs = [pipeline.get_outputs(**kwargs) for pipeline in self] # pylint: disable=redefined-outer-name
    203         if return_sum:
    204             if len(outputs) > 1:

/afs/ifh.de/group/amanda/scratch/blotsd/Analysis/Oscillations/fork_pisa3/pisa/core/distribution_maker.py in <listcomp>(.0)
    200 
    201         """
--> 202         outputs = [pipeline.get_outputs(**kwargs) for pipeline in self] # pylint: disable=redefined-outer-name
    203         if return_sum:
    204             if len(outputs) > 1:

/afs/ifh.de/group/amanda/scratch/blotsd/Analysis/Oscillations/fork_pisa3/pisa/utils/profiler.py in profiled_func(*args, **kwargs)
    123         try:
    124             start_t = time()
--> 125             return func(*args, **kwargs)
    126         finally:
    127             end_t = time()

/afs/ifh.de/group/amanda/scratch/blotsd/Analysis/Oscillations/fork_pisa3/pisa/core/pipeline.py in get_outputs(self, inputs, idx, return_intermediate)
    356             try:
    357                 logging.trace(">>> BEGIN: {}.run(...)".format(name))
--> 358                 outputs = stage.run(inputs=inputs) # pylint: disable=redefined-outer-name
    359                 if return_intermediate:
    360                     if outputs is None:  # e.g. for PISA pi

/afs/ifh.de/group/amanda/scratch/blotsd/Analysis/Oscillations/fork_pisa3/pisa/core/pi_stage.py in run(self, inputs)
    261         if not inputs is None:
    262             raise ValueError("PISA pi requires there not be any inputs.")
--> 263         self.compute()
    264         self.apply()
    265         return None

/afs/ifh.de/group/amanda/scratch/blotsd/Analysis/Oscillations/fork_pisa3/pisa/utils/profiler.py in profiled_func(*args, **kwargs)
    123         try:
    124             start_t = time()
--> 125             return func(*args, **kwargs)
    126         finally:
    127             end_t = time()

/afs/ifh.de/group/amanda/scratch/blotsd/Analysis/Oscillations/fork_pisa3/pisa/core/pi_stage.py in compute(self)
    219             for container in self.data:
    220                 for key in self.output_calc_keys:
--> 221                     container.binned_to_array(key)
    222 
    223     def compute_function(self):

/afs/ifh.de/group/amanda/scratch/blotsd/Analysis/Oscillations/fork_pisa3/pisa/core/container.py in binned_to_array(self, key)
    399                 return
    400             else:
--> 401                 raise ValueError('Key `%s` does not exist in container `%s`'%(key, self.name))
    402         logging.debug('Transforming %s binned to array data'%(key))
    403         sample = [self.array_data[n] for n in binning.names]

ValueError: Key `n` does not exist in container `nue_cc`

Reply:
here it says now there is no key "n"....? This is a different error to what you posted above

Reply:
yeah I restarted my notebook to be sure I was trying on a clean slate, but now the error message changes to what is above.